common/transaction/priority_queue.go

```diff
@@ -0,0 +1,51 @@
+package transaction
+
+// PriorityQueue implements a priority queue using a double linked list
+type PriorityQueue struct {
+	head *node
+}
+
+type node struct {
+	data   *ValidTransaction
+	parent *node
+	child  *node
+}
+
+// Pop removes the head of the queue and returns it
+func (q *PriorityQueue) Pop() *ValidTransaction {
+	head := q.head
+	q.head = head.child
+	return head.data
+}
+
+// Insert traverses the list and places a valid transaction with priority p directly before the
+// first node with priority p-1. If there are other nodes with priority p, the new node is placed
+// behind them.
+func (q *PriorityQueue) Insert(vt *ValidTransaction) {
+	curr := q.head
+	if curr == nil {
+		q.head = &node{data: vt}
+		return
+	}
+
+	for {
```

Comment:
Infinite loop possible. I think this can be written as `for ; curr.child != nil; curr = curr.child`

Reply:
Also -- I don't think this works for the case where the queue is [`2`] and you insert `1`

Reply:
I have changed the loop to not be an infinite loop anymore and changed the logic so that if the curr's child is nil, then we just insert the node